crates/uv-resolver/src/pubgrub/package.rs

```diff
@@ -117,6 +121,72 @@ impl PubGrubPackage {
         }
     }
 
+    /// Modifies this package in place such that it is associated with the
+    /// given markers by intersecting them any pre-existing markers.
+    ///
+    /// That is, this causes the package to only be applicable to marker
+    /// environments corresponding to the intersection of what it previously
+    /// matched and the given markers.
+    ///
+    /// This is useful when one needs to propagate markers from a fork to each
+    /// of its constituent dependencies. This is necessary because within a
+    /// fork, the resolver makes decisions based on the markers that created
+    /// that fork. While in many cases these decisions are universal, some may
+    /// not be! And so the markers from the fork must be propagated out to the
+    /// individual packages.
+    pub(crate) fn and_markers(&mut self, fork_marker: &MarkerTree) {
+        // We do a little dance here to pluck out as much existing
+        // information from this package as we can to avoid allocs. It
+        // is possible that the `Arc::make_mut` will do a deep clone
+        // here, thereby defeating our efforts, but I think it is
+        // likely that in most cases it will not.
+        match Arc::make_mut(&mut self.0) {
+            PubGrubPackageInner::Root(_) | PubGrubPackageInner::Python(_) => {}
+            // In this case, we may or may not already have a marker.
+            // If we don't, then this implies we will, and thus we may
+            // need to switch `Package` to some other representation.
+            PubGrubPackageInner::Package {
+                ref mut name,
+                ref mut extra,
+                ref dev,
+                ref mut marker,
+            } => {
+                // FIXME: It's not quite clear what to do in this case. It's
+                // also not totally clear whether it can actually happen
+                // either.
+                if dev.is_some() {
+                    return;
+                }
```

Comment:
I am unsure of what to do in this case. I haven't given it too much thought yet.